.github/workflows/update-gpuci.yml

```diff
@@ -13,19 +13,23 @@ jobs:
     steps:
       - uses: actions/checkout@v4
 
-      - name: Parse current axis YAML
+      - name: Fetch current RAPIDS versions from build matrix
+        id: parse_yaml
         uses: the-coding-turtle/ga-yaml-parser@v0.1.2
         with:
-          file: ci/axis/dask.yaml
-          export_to_envs: true
-          return_to_outputs: false
+          file: .github/workflows/build.yml
+          export_to_envs: false
+          return_to_outputs: true
 
       - name: Get old RAPIDS / UCX-Py versions
+        env:
+          OLD_RAPIDS_VER: ${{ steps.parse_yaml.outputs.jobs_docker_strategy_matrix_rapids_0 }}
+          RAPIDS_VER: ${{ steps.parse_yaml.outputs.jobs_docker_strategy_matrix_rapids_1 }}
```

Comment:
I'm uncomfortable with implicitly depending on the ordering of the `matrix.rapids` entry here. I could see someone changing `["24.10", "24.12"]` to `["25.02", "24.12"]` in a PR and not realizing that that was breaking some of the assumptions here.

I also think it's wasteful (and a source of unnecessary risk of failed builds due to network issues) to have all these calls to `https://version.gpuci.io/rapids`. The response from that endpoint should change only when the RAPIDS version changes. And if you're only tracking 2 RAPIDS versions, I think it'd be simpler to just hard-code the corresponding `ucx-py` version.

Would you consider something like the following over in `build.yml`? At the top level of the workflow:

```yaml
env:
  OLD_RAPIDS_VER: "24.10"
  OLD_UCX_PY_VER: "0.40"
  RAPIDS_VER: "24.12"
  UCX_PY_VER: "0.41"

jobs:
  docker:
    runs-on: ubuntu-latest

    strategy:
      fail-fast: false
      matrix:
        cuda: ["11.8.0"]
        python: ["3.10", "3.11", "3.12"]
        linux: ["ubuntu20.04"]
        rapids: [${{ env.OLD_RAPIDS_VER }}, ${{ env.NEW_RAPIDS_VER }}]
```

And then passing those values as outputs through to whatever needs it in this `update-gpuci.yml`? That'd remove that source of network calls, and remove this implicit reliance on the ordering of the matrix entries for correctness.

Reply:
I think that'd also allow you to get rid of all this other machinery:

https://github.com/rapidsai/dask-build-environment/blob/89770692cee67f261d0a5cdd5f5b28ef82342450/.github/workflows/update-gpuci.yml#L38-L44

There's exactly 1 `ucx-py` version per RAPIDS release and these builds seem to assume they'll only run for exactly 2 RAPIDS versions... seems like a great opportunity to replace some complicated and custom machinery with a little bit of hard-coding that'll be updated infrequently.

Reply:
Yeah I definitely think the ordering concern alone is sufficient to hardcode these values somewhere other than the build matrix itself, shouldn't be too difficult to refactor the builds to rely on an internal version mapping instead of `https://version.gpuci.io/rapids`

> I think that'd also allow you to get rid of all this other machinery

Were you thinking of replacing this step with a request to `https://version.gpuci.io/rapids` to fetch the new UCX-Py version instead? That could work, IIRC I opted to do it this way because cuDF and UCX-Py nightlies became available at different times in the RAPIDS release process, and so this ensured that both were available before opening a PR that would inevitably fail at the conda solve step (though this may be less of a problem now with PR checks running to make PRs like that visibly "blocked" from merging)

Reply:
Another minor annoyance here I didn't foresee - it looks like the `env` context is not available within the `matrix` block, so the example you shared above triggers a syntax error:

https://github.com/rapidsai/dask-build-environment/actions/runs/11114625627/workflow

Reply:
Was able to get around this by using some hardcoded representative values in the build matrix, that we then later manually map to the hardcoded RAPIDS/ucx-py versions in https://github.com/rapidsai/dask-build-environment/pull/105/commits/0af7f24e726b2170de32201e1518fc3c5d26449a

Reply:
> *were you thinking of replacing this step*

I looked back through this workflow more closely to be sure I understood what it was doing. I didn't understand that it's job was to trigger new PRs here, so some of my earlier comments about it might not make sense.

Am I right that this `update-gpuci` workflow's role is "open a PR here in `dask-build-environment` whenever a new RAPIDS release or `ucx-py` release is detected?".

https://github.com/rapidsai/dask-build-environment/blob/89770692cee67f261d0a5cdd5f5b28ef82342450/.github/workflows/update-gpuci.yml#L86-L98

Like #104?

If so... I'd just remove this entire `update-gpuci.yml` workflow. Maybe things were different back when it was introduced in #10, but today the amount of manual work is so small here and happens so infrequently, I just don't think all this complexity in GitHub Actions is worth it.

There's a new RAPIDS release once every 2 months, and every RAPIDS release corresponds to exactly one `ucx-py` release.

Specifically, I'm proposing:

* remove `update-gpuci.yml` workflow completely
* ask RAPIDS ops if updating this repo can be a part of the RAPIDS release checklist
   - *(optional): automate those updates via a `ci/release/update-version.sh`, like most other RAPIDS repos do, e.g. https://github.com/rapidsai/docker/blob/branch-24.10/ci/release/update-version.sh*

If you want to keep this automated, then ignore my other comments about `update-gpuci` and leave the network calls to get those versions as-is.

Reply:
> *looks like the env context is not available within the matrix block*

blegh ok, sorry for misleading you there. I like the solution you came up with!

Reply:
> Am I right that this update-gpuci workflow's role is "open a PR here in dask-build-environment whenever a new RAPIDS release or ucx-py release is detected?".

Pretty much! Only thing I'd add is that it's specifically when a new _cuDF_ release _and_ ucx-py release are detected, not if it's just or the other.

> ask RAPIDS ops if updating this repo can be a part of the RAPIDS release checklist

Definitely a strong +1 on adding something like this to a checklist to increase the visibility of the ops work around Dask GPU CI, worth noting that there are similar workflows in the respective Dask repos to automate this version bump too:

- https://github.com/dask/dask/blob/main/.github/workflows/update-gpuci.yml
- https://github.com/dask/distributed/blob/main/.github/workflows/update-gpuci.yaml

Which reflect other steps that could probably be done manually as part of a checklist - maybe we should file an issue here to track this switch?

Reply:
Sure! Up to you and the other Dask folks how you want to handle it. Also fine to tell me "nah we'll keep this as-is", I don't feel that strongly about it. Just right now, in this moment, this feels like a lot of complexity to automate something that changes very infrequently.